### DISCUSSION TITLE
cypress connect with Microsoft SQL but error occured

### DISCUSSION THREAD

**Discussion by Habib0** [cypress-io/cypress#27350] [Questions and Help]
cypress.json file
{
  "env": {
    "db": {
      "user": "habib007",
      "password": "123",
      "server": "DESKTOP-KFJKDD2",
      "database": "testdb",
      "options": {
        "trustedConnection": false
      }
    }
  }
}

index.js plugin file

const sql = require("mssql");

async function queryTestDb(query, config) {
  try {
    await sql.connect(config.env.db);
    const result = await sql.query(query);
    return result.recordset;
  } catch (error) {
    throw error;
  } finally {
    sql.close();
  }
}

module.exports = (on, config) => {
  on('task', {
    queryDb: (query) => {
      return queryTestDb(query, config);
    },
  });
};

intergration file
describe('Database Testing', () => {
    it('should retrieve data from the database', () => {

      cy.task('queryDb', 'SELECT * FROM users').then((results) => {
        // Use the retrieved data for assertions or further actions
        const newdata = results;
        console.log(newdata)
        expect(results.length).to.be.greaterThan(0);
      });
    });
  });

package.json file:

 "devDependencies": {
    "@babel/core": "^7.22.9",
    "@babel/preset-env": "^7.22.9",
    "babel-loader": "^9.1.3",
    "cypress-sql-server": "^1.0.0",
    "mssql": "^9.1.1",
    "mysql": "^2.18.1",
    "tedious": "^16.2.0"
  }
}

but error occured:
task
queryDb, SELECT * FROM users
CypressError
cy.task('queryDb') failed with the following error:

Failed to connect to DESKTOP-KFJKDD2:1433 - Could not connect (sequence).

![image](https://github.com/cypress-io/cypress/assets/31403162/6fd1e30b-2a47-484a-995b-4e33ffe9d746)

**Comment by portyu9** [comment]
Hi @Habib0 

`cy.task()` is working here. The failure happens **inside the Node task**, when `mssql`/Tedious attempts to open the SQL Server connection:

```text
Failed to connect to DESKTOP-KFJKDD2:1433 - Could not connect
```

So this is a SQL Server TCP connectivity/configuration problem rather than Cypress being unable to execute the database task.

Check these in order:

1. Confirm which TCP port that SQL Server instance is actually listening on.
2. Make sure TCP/IP is enabled for the instance in SQL Server Configuration Manager.
3. Restart SQL Server after changing its TCP configuration.
4. Make sure Windows Firewall permits the configured TCP port.
5. Test that port from the same machine/account running Cypress.

`node-mssql` defaults to port `1433` for a normal/default instance. A **named instance** (including many SQL Express installations) commonly uses a dynamic port instead. In that case configure the actual port, or configure the named instance appropriately rather than assuming `1433`.

For example, once you know the port:

```js
const db = {
  user: '...',
  password: '...',
  server: 'DESKTOP-KFJKDD2',
  database: 'testdb',
  port: 1433,
}
```

Also note that the shown `cypress.json` + `cypress/plugins/index.js` structure is legacy Cypress syntax. In current Cypress, register the same `task` from `e2e.setupNodeEvents()` in `cypress.config.js/ts`.

That migration is separate from this error, though—the immediate failure is that the SQL client cannot establish the TCP connection.

Please mark this answer as accepted if it helped, thank you.
